File: WordPattern_LC_290.java
/*
 * time Complexity - O(n)
 * Space complexity - O(n) + O (n) = 0(n) hashmap 
 * maintain hashmap with index value of char and word
 */

package Super30;

import java.util.HashMap;

public class WordPattern {

	public static void main(String[] args) {
		String pattern = "ccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccdd";
		String s = "s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s s t t";
		
		System.out.println(wordPattern(pattern, s));

	}
	
	
	public static boolean wordPattern(String pattern, String s) {
        HashMap map = new HashMap();
        
        String[] words = s.split(" ");
        
        if(pattern.length() != words.length)
            return false;
        
        for(int i = 0;  i < pattern.length(); i++){
            char c = pattern.charAt(i);
            String w = words[i];
            
            if(!map.containsKey(c)){
                map.put(c, i);            
            }
            
            if(!map.containsKey(w)){
                map.put(w, i);
            }
             
            int a = (int) map.get(c);
            int b = (int) map.get(w);
            if(a != b)
                return false;
        }
        return true;
    }
}

File: wordPattern-LC 290.py
# Time Complexityn= O(n)
# Space Complexity = O(1)


class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        
        # 2 hashmaps solutions. Using dictionaries in python
        
        # Storing the corresponding strings in a list, separated by " "
        strList = s.split(" ")
        
        if (len(pattern) != len(strList)):
            return False
        
        patternDict = {}
        sDict = {}
        
        # Iteration of characters
        for i in range(0, len(pattern)):
            pChar = pattern[i]
            sStr = strList[i]
            # Character already present in pattern Dictionary
            if pChar in patternDict:
                # Check if pChar is mapped to the same string
                if patternDict.get(pChar) != sStr:
                    return False
            # Add new character to dictionary
            else:
                patternDict.update({pChar: sStr})
            
            # Character already present in s Dictionary
            if sStr in sDict:
                # Check if string is mapped to the same pChar
                if sDict.get(sStr) != pChar:
                    return False
            # Add new character to dictionary
            else:
                sDict.update({sStr: pChar})
        
        return True
        
